In [36]:
import pandas as pd

In [37]:
data=pd.read_csv('processed_local_large_well_3_1c_dfn.csv', float_precision='round_trip')

In [38]:
data.describe()

Id             X             Y             Z         dX  \
count  3.920000e+06  3.920000e+06  3.920000e+06  3.920000e+06  3920000.0   
mean   1.960000e+06  1.763700e+03  1.228970e+03 -4.750000e+02       10.0   
std    1.131607e+06  4.041349e+02  4.041349e+02  5.773431e+02        0.0   
min    0.000000e+00  1.068700e+03  5.339700e+02 -1.470000e+03       10.0   
25%    9.799998e+05  1.416200e+03  8.814700e+02 -9.725000e+02       10.0   
50%    1.960000e+06  1.763700e+03  1.228970e+03 -4.750000e+02       10.0   
75%    2.939999e+06  2.111200e+03  1.576470e+03  2.250000e+01       10.0   
max    3.919999e+06  2.458700e+03  1.923970e+03  5.200000e+02       10.0   

              dY         dZ      Porosity           kii           kjj  \
count  3920000.0  3920000.0  3.920000e+06  3.920000e+06  3.920000e+06   
mean        10.0       10.0  7.011576e-03  3.595809e-16  3.589851e-16   
std          0.0        0.0  1.558229e-02  2.288301e-15  2.288239e-15   
min         10.0       10.0  3.316340e-04  1.023090e-18  1.159695e-18   
25%         10.0       10.0  4.000981e-03  3.064824e-17  2.994944e-17   
50%         10.0       10.0  4.839916e-03  4.127572e-17  4.073644e-17   
75%         10.0       10.0  5.758113e-03  5.421798e-17  5.400832e-17   
max         10.0       10.0  1.200000e-01  1.700000e-14  1.700000e-14   

                kkk         block  
count  3.920000e+06  3.920000e+06  
mean   3.639264e-16  1.981456e+00  
std    2.288104e-15  1.349083e-01  
min    1.248545e-18  1.000000e+00  
25%    3.360730e-17  2.000000e+00  
50%    4.511092e-17  2.000000e+00  
75%    5.907400e-17  2.000000e+00  
max    1.700000e-14  2.000000e+00

In [39]:
min (data.X), max (data.X), data['X'].median()

(1068.7, 2458.7, 1763.7)

In [40]:
min (data.Y), max (data.Y), data['Y'].median()

(533.97, 1923.97, 1228.97)

In [41]:
min (data.Z), max (data.Z), data['Z'].median()

(-1470, 520, -475.0)

now make a copy for each simulation case, and translate the mesh

start with the 3-1c trajectory case

In [42]:
large_well_3_1c = data.copy(deep=True)

need to manually calculate the translation distance based on injection location and bounds of the new/refine mesh versus the full domain mesh

In [43]:
large_well_3_1c.X = large_well_3_1c.X+36.3

In [44]:
large_well_3_1c.Y = large_well_3_1c.Y+56.03

In [10]:
large_well_3_1c.Z = large_well_3_1c.Z-620

now verify translation was correct, and also chck the bounds to make sure still within the full domain

In [45]:
min (large_well_3_1c.X), max (large_well_3_1c.X), large_well_3_1c['X'].median()

(1105.0, 2495.0, 1800.0)

In [46]:
min (large_well_3_1c.Y), max (large_well_3_1c.Y), large_well_3_1c['Y'].median()

(590.0, 1980.0, 1285.0)

In [13]:
min (large_well_3_1c.Z), max (large_well_3_1c.Z), large_well_3_1c['Z'].median()

(-1740, 250, -745.0)

X and Z is out of bounds, min value in full domain is -1475 (for the nodes).  min for 10m cell would be -1470

fix Z

In [14]:
large_well_3_1c.Z = large_well_3_1c.Z+270

In [15]:
min (large_well_3_1c.Z), max (large_well_3_1c.Z), large_well_3_1c['Z'].median()

(-1470, 520, -475.0)

fix X

In [16]:
large_well_3_1c.X = large_well_3_1c.X-125

In [17]:
min (large_well_3_1c.X), max (large_well_3_1c.X), large_well_3_1c['X'].median()

(1105.0, 2495.0, 1800.0)

now all is good for the 3-1c case.  BUT...because the file is so big, it can't be edited in Excel, so need to add some additional columns to the file now, and then later add the 10 row header that is usually written by Leapfrog

need to add column (1) for Id, columns 5, 6, and 7 for dX, dY, and dZ

first add the "Id"

In [8]:
large_well_3_1c.insert(loc=0, column='Id', value=large_well_3_1c.index)

now add the dX, dY, and dZ

In [9]:
large_well_3_1c.insert(loc=4, column='dX', value=10)

In [10]:
large_well_3_1c.insert(loc=5, column='dY', value=10)

In [11]:
large_well_3_1c.insert(loc=6, column='dZ', value=10)

now check the file

In [12]:
large_well_3_1c.head()

Id       X        Y    Z  dX  dY  dZ  Porosity  Compressibility  \
0   0  1068.7  1923.97  520  10  10  10  0.001836     1.176111e-08   
1   1  1068.7  1923.97  510  10  10  10  0.001661     1.122519e-08   
2   2  1068.7  1923.97  500  10  10  10  0.001703     1.089720e-08   
3   3  1068.7  1923.97  490  10  10  10  0.001489     1.010419e-08   
4   4  1068.7  1923.97  480  10  10  10  0.002365     1.572026e-08   

            kii           kjj           kkk  
0  8.296096e-18  1.454657e-17  1.643608e-17  
1  5.336351e-18  1.292858e-17  1.789974e-17  
2  4.304728e-18  1.992727e-17  2.258996e-17  
3  9.033887e-18  1.493932e-17  1.425128e-17  
4  1.105784e-17  1.891407e-17  1.917711e-17

now sort so values are in correct order for em2ex

In [13]:
a = large_well_3_1c.sort_values(["Z","Y","X"], ascending = [True, True, True])

In [14]:
a.head()

Id       X       Y     Z  dX  dY  dZ  Porosity  Compressibility  \
27999    27999  1068.7  533.97 -1470  10  10  10  0.001921     1.095691e-08   
55999    55999  1078.7  533.97 -1470  10  10  10  0.001408     9.427024e-09   
83999    83999  1088.7  533.97 -1470  10  10  10  0.001895     1.090580e-08   
111999  111999  1098.7  533.97 -1470  10  10  10  0.002060     1.269264e-08   
139999  139999  1108.7  533.97 -1470  10  10  10  0.002268     1.545065e-08   

                 kii           kjj           kkk  
27999   8.004419e-18  1.272226e-17  1.039621e-17  
55999   1.363619e-17  1.328845e-17  8.272244e-18  
83999   9.412358e-18  1.801207e-17  6.617931e-18  
111999  1.722060e-17  1.749362e-17  2.737495e-17  
139999  2.306947e-17  2.791417e-17  6.828935e-17

now we have the perm and proposity all set up, but if the mesh has both granite and sediment, we need to all a column for the block number, and re-assign the perm and porosity for the sediments

first export the block model from Leapfrog, then run global2local.py to get it into local coords, then read it in here

In [15]:
GMdata=pd.read_csv('local_large_3-1c_Block Model_cell.csv', float_precision='round_trip')

In [16]:
b = GMdata.sort_values(["Z","Y","X"], ascending = [True, True, True])

In [17]:
b.head()

Id        X       Y     Z  dX  dY  dZ         GM  pressure_MPa  \
0   0  1067.91  531.78 -1470  10  10  10  Granitiod     27.538509   
1   1  1077.91  531.78 -1470  10  10  10  Granitiod     27.537451   
2   2  1087.91  531.78 -1470  10  10  10  Granitiod     27.536406   
3   3  1097.91  531.78 -1470  10  10  10  Granitiod     27.535365   
4   4  1107.91  531.78 -1470  10  10  10  Granitiod     27.534316   

   sigma_h_max_MPa  sigma_h_min_MPa  Sigma_V_MPa  temperature_C  
0        54.871706        46.978296    79.795385     245.434904  
1        54.876142        46.973138    79.806631     245.681018  
2        54.880795        46.968362    79.818372     245.928550  
3        54.885422        46.963590    79.830094     246.177700  
4        54.889690        46.958394    79.841148     246.428922

sediments = block 1, grnit = block 2

In [18]:
b['block'] = 0

In [19]:
b.loc[b.GM == 'Granitiod', 'block'] = 2

In [20]:
b.GM.unique()

array(['Granitiod', 'Valley Fill'], dtype=object)

In [21]:
b.loc[b.GM == 'Valley Fill', 'block'] = 1

In [22]:
b.block.unique()

array([2, 1])

In [23]:
b.head()

Id        X       Y     Z  dX  dY  dZ         GM  pressure_MPa  \
0   0  1067.91  531.78 -1470  10  10  10  Granitiod     27.538509   
1   1  1077.91  531.78 -1470  10  10  10  Granitiod     27.537451   
2   2  1087.91  531.78 -1470  10  10  10  Granitiod     27.536406   
3   3  1097.91  531.78 -1470  10  10  10  Granitiod     27.535365   
4   4  1107.91  531.78 -1470  10  10  10  Granitiod     27.534316   

   sigma_h_max_MPa  sigma_h_min_MPa  Sigma_V_MPa  temperature_C  block  
0        54.871706        46.978296    79.795385     245.434904      2  
1        54.876142        46.973138    79.806631     245.681018      2  
2        54.880795        46.968362    79.818372     245.928550      2  
3        54.885422        46.963590    79.830094     246.177700      2  
4        54.889690        46.958394    79.841148     246.428922      2

now we have a sorted column that has block numbers, need to add it to the dataframe with the permeability and porosity

In [24]:
b.index = a.index

In [25]:
a['block'] = 0

In [26]:
a['block'] = b['block'].values

In [27]:
a.head()

Id       X       Y     Z  dX  dY  dZ  Porosity  Compressibility  \
27999    27999  1068.7  533.97 -1470  10  10  10  0.001921     1.095691e-08   
55999    55999  1078.7  533.97 -1470  10  10  10  0.001408     9.427024e-09   
83999    83999  1088.7  533.97 -1470  10  10  10  0.001895     1.090580e-08   
111999  111999  1098.7  533.97 -1470  10  10  10  0.002060     1.269264e-08   
139999  139999  1108.7  533.97 -1470  10  10  10  0.002268     1.545065e-08   

                 kii           kjj           kkk  block  
27999   8.004419e-18  1.272226e-17  1.039621e-17      2  
55999   1.363619e-17  1.328845e-17  8.272244e-18      2  
83999   9.412358e-18  1.801207e-17  6.617931e-18      2  
111999  1.722060e-17  1.749362e-17  2.737495e-17      2  
139999  2.306947e-17  2.791417e-17  6.828935e-17      2

In [28]:
a.loc[a.block == 1, 'kii'] = 1.7e-14

In [29]:
a.loc[a.block == 1, 'kjj'] = 1.7e-14

In [30]:
a.loc[a.block == 1, 'kkk'] = 1.7e-14

In [31]:
a.loc[a.block == 1, 'Porosity'] = 0.12

In [32]:
del a['Compressibility']

In [33]:
a.describe()

Id             X             Y             Z         dX  \
count  3.920000e+06  3.920000e+06  3.920000e+06  3.920000e+06  3920000.0   
mean   1.960000e+06  1.763700e+03  1.228970e+03 -4.750000e+02       10.0   
std    1.131607e+06  4.041349e+02  4.041349e+02  5.773431e+02        0.0   
min    0.000000e+00  1.068700e+03  5.339700e+02 -1.470000e+03       10.0   
25%    9.799998e+05  1.416200e+03  8.814700e+02 -9.725000e+02       10.0   
50%    1.960000e+06  1.763700e+03  1.228970e+03 -4.750000e+02       10.0   
75%    2.939999e+06  2.111200e+03  1.576470e+03  2.250000e+01       10.0   
max    3.919999e+06  2.458700e+03  1.923970e+03  5.200000e+02       10.0   

              dY         dZ      Porosity           kii           kjj  \
count  3920000.0  3920000.0  3.920000e+06  3.920000e+06  3.920000e+06   
mean        10.0       10.0  7.011576e-03  3.595809e-16  3.589851e-16   
std          0.0        0.0  1.558229e-02  2.288301e-15  2.288239e-15   
min         10.0       10.0  3.316340e-04  1.023090e-18  1.159695e-18   
25%         10.0       10.0  4.000981e-03  3.064824e-17  2.994944e-17   
50%         10.0       10.0  4.839916e-03  4.127572e-17  4.073644e-17   
75%         10.0       10.0  5.758113e-03  5.421798e-17  5.400832e-17   
max         10.0       10.0  1.200000e-01  1.700000e-14  1.700000e-14   

                kkk         block  
count  3.920000e+06  3.920000e+06  
mean   3.639264e-16  1.981456e+00  
std    2.288104e-15  1.349083e-01  
min    1.248545e-18  1.000000e+00  
25%    3.360730e-17  2.000000e+00  
50%    4.511092e-17  2.000000e+00  
75%    5.907400e-17  2.000000e+00  
max    1.700000e-14  2.000000e+00

now can write the data to a CSV file.

In [34]:
a.to_csv("processed_local_large_well_3_1c_dfn.csv", encoding='utf-8', index=False)

In [35]:
a.describe().to_csv("processed_local_large_well_3_1c_dfn_Summary.csv")

now need to read into leapfrog and verify all the mpaaings